### PR TITLE
[feat]: [CDS-15080]: PDC SSH Instance Sync

### PR DESCRIPTION
**Pull request by arvind-choudhary-h**
Important: To change perpetual task frequency from 1 to 10 minutes

You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)

### CODE REVIEW

**Comment by arvind-choudhary-h** [comment]
trigger messagecheck

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 14 of 23 files at r1, 11 of 12 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java, line 202 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK3-q22jH8oFvhPBbN:-MxK3-q22jH8oFvhPBbO:bqxkwam) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java#L202)):*
> ```Java
>     Map<String, Boolean> reachableMap = checkReachability(hostNames, settingAttribute, encryptedDataDetails);
> 
>     Set<String> instancesToRemove = instances.stream()
> ```

We are saving in DB only reachable instances, right? 
___
*[400-rest/src/main/java/software/wings/utils/HostValidationServiceImpl.java, line 133 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK4g440lWPWNHFmcc9:-MxK4g440lWPWNHFmccA:b-yhtyiu) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/400-rest/src/main/java/software/wings/utils/HostValidationServiceImpl.java#L133)):*
> ```Java
>       result.add(HostReachabilityInfo.builder()
>                      .hostName(host)
>                      .reachable(SocketConnectivityCapabilityCheck.connectableHost(host, portf))
> ```

We are checking only socket connection, do we need validation of credentials?
___
*[956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java, line 1 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK1ZTw8JZdYNBA4Cl2:-MxK1ZTw8JZdYNBA4Cl3:bo6plka) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L1)):*
> ```Java
> /*
> ```

Do we have a new FF?

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798#-:-MxK4xgG0pYy23Ar7Zje:bnfp4nl)* status: :shipit: complete! ✓

**Comment by arvind-choudhary-h** [comment]
___
*[400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java, line 202 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK3-q22jH8oFvhPBbN:-MxK6-85B__Jc3j6voFI:bs6ompv) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java#L202)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

We are saving in DB only reachable instances, right? 
</blockquote></details>

Yes @ivan-83. It's being done in software.wings.service.impl.instance.InstanceHelper#extractInstanceOrDeploymentInfoBaseOnType post deployment

**Comment by arvind-choudhary-h** [comment]
___
*[400-rest/src/main/java/software/wings/utils/HostValidationServiceImpl.java, line 133 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK4g440lWPWNHFmcc9:-MxK6JLjC9qZuG0sw-Ua:b-g4oxhq) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/400-rest/src/main/java/software/wings/utils/HostValidationServiceImpl.java#L133)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

We are checking only socket connection, do we need validation of credentials?
</blockquote></details>

No. Socket connectivity was decided during design phase as every 10 minutes, it will be a lot.

**Comment by arvind-choudhary-h** [comment]
___
*[956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java, line 1 at r3](https://reviewable.io/reviews/harness/harness-core/30798#-MxK1ZTw8JZdYNBA4Cl2:-MxK6THz9jUihAlz3x5s:b-qaub5t) ([raw file](https://github.com/harness/harness-core/blob/f47fd648aa2fb0ed27a46bad56abab2de14cd8ce/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L1)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Do we have a new FF?
</blockquote></details>

Basically, we'll use 2 FF with this:
1. PDC_PERPETUAL_TASK (For enabling PT)
2. STOP_INSTANCE_SYNC_VIA_ITERATOR_FOR_PDC_DEPLOYMENTS (Default iterator is enabled. To disable iterator Fk, we'll need to enable this too)

**Review by ivan-83 — COMMENTED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: 3. unresolved discussions. 0 of 1 approvals obtained.

**Review by ivan-83 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: :shipit: complete! ✓

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 11 of 23 files at r1, 9 of 12 files at r2, 1 of 2 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: :shipit: complete! ✓

___
*[260-delegate/src/main/java/software/wings/delegatetasks/HostValidationTask.java, line 76 at r4](https://reviewable.io/reviews/harness/harness-core/30798#-MxOKyDkCV5ClIcRQnnB:-MxOKyDkCV5ClIcRQnnC:bsndjo8) ([raw file](https://github.com/harness/harness-core/blob/5ffbbd418a2136d0ed090a7c72646c072070216d/260-delegate/src/main/java/software/wings/delegatetasks/HostValidationTask.java#L76)):*
> ```Java
>   }
> 
>   private DelegateResponseData getTaskExecutionResponseData(HostValidationTaskParameters hostValidationTaskParameters) {
> ```

We should verify Test connection of PDC infra def and Test Connection while creating ssh/winrm connection attributes
___
*[400-rest/src/main/java/io/harness/perpetualtask/PdcPTClientParams.java, line 21 at r4](https://reviewable.io/reviews/harness/harness-core/30798#-MxOAeJj5zSyQl0TxsR0:-MxOAeJk0lhOYKDM-CmD:beaetmf) ([raw file](https://github.com/harness/harness-core/blob/5ffbbd418a2136d0ed090a7c72646c072070216d/400-rest/src/main/java/io/harness/perpetualtask/PdcPTClientParams.java#L21)):*
> ```Java
> @OwnedBy(CDP)
> public class PdcPTClientParams implements PerpetualTaskClientParams {
>   String inframappingId;
> ```

fields can be private ?
___
*[400-rest/src/main/java/software/wings/service/impl/instance/AzureInstanceHandler.java, line 89 at r4](https://reviewable.io/reviews/harness/harness-core/30798#-MxOGniY95LpmCiAU4yK:-MxOGniZ5ZUy13WHEK2m:baw3ueh) ([raw file](https://github.com/harness/harness-core/blob/5ffbbd418a2136d0ed090a7c72646c072070216d/400-rest/src/main/java/software/wings/service/impl/instance/AzureInstanceHandler.java#L89)):*
> ```Java
>       }
> 
>       if (instanceInfo instanceof AzureVMSSInstanceInfo) {
> ```

nit: else if ?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 1 of 23 files at r1, 1 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: :shipit: complete! ✓

___
*[400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java, line 169 at r4](https://reviewable.io/reviews/harness/harness-core/30798#-MxRmSC9740buIEb9ALE:-MxRmSCJ-wKIUyoq9aK1:b-cwclav) ([raw file](https://github.com/harness/harness-core/blob/5ffbbd418a2136d0ed090a7c72646c072070216d/400-rest/src/main/java/software/wings/service/impl/instance/PdcInstanceHandler.java#L169)):*
> ```Java
>   public void processInstanceSyncResponseFromPerpetualTask(
>       InfrastructureMapping infrastructureMapping, DelegateResponseData response) {
>     syncInstances(infrastructureMapping.getAppId(), infrastructureMapping.getUuid(), InstanceSyncFlow.PERPETUAL_TASK);
> ```

are we not using the response from the perpetual task here ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 23 files at r1, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: 1 change requests. 0 of 1 approvals obtained.

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: 1 change requests. 0 of 1 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30798)* status: :shipit: complete! ✓

**Comment by arvind-choudhary-h** [comment]
trigger tin0

**Comment by arvind-choudhary-h** [comment]
trigger tin1

**Comment by arvind-choudhary-h** [comment]
trigger ti2

**Comment by arvind-choudhary-h** [comment]
trigger ti0,ti1,ti3,ti4
